godot-macros/src/docs.rs

```diff
@@ -0,0 +1,224 @@
+/*
+ * Copyright (c) godot-rust; Bromeon and contributors.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+mod markdown_converter;
+
+use crate::class::{ConstDefinition, Field, FuncDefinition, SignalDefinition};
+use proc_macro2::{Ident, TokenStream};
+use quote::{quote, ToTokens};
+use venial::*;
+
+pub fn make_definition_docs(
+    base: String,
+    description: &[Attribute],
+    members: &[Field],
+) -> TokenStream {
+    (|| {
+        let desc = make_docs_from_attributes(description)?;
+        let members = members
+            .into_iter()
+            .filter(|x| x.var.is_some() | x.export.is_some())
+            .map(member)
+            .collect::<Option<String>>()?;
+        Some(quote! {
+            docs: ::godot::docs::StructDocs {
+                base: #base,
+                description: #desc,
+                members: #members,
+            }.into()
+        })
+    })()
+    .unwrap_or(quote! { docs: None })
+}
+
+pub fn make_inherent_impl_docs(
+    functions: &[FuncDefinition],
+    constants: &[ConstDefinition],
+    signals: &[SignalDefinition],
+) -> TokenStream {
+    (|| {
+        let methods = functions
+            .iter()
+            .map(make_method_docs)
+            .collect::<Option<String>>()?;
+        let signals = signals
+            .iter()
+            .map(make_signal_docs)
+            .collect::<Option<String>>()?;
+        let constants = constants
+            .iter()
+            .map(|ConstDefinition { raw_constant: x }| x)
+            .map(constant)
+            .collect::<Option<String>>()?;
+        Some(quote! {
+            docs: ::godot::docs::InherentImplDocs {
+                methods: #methods,
+                signals: #signals,
+                constants: #constants,
+            }.into()
+        })
+    })()
+    .unwrap_or(quote! { docs: None })
```

Comment:
Could you please prefer more procedural and easy-to-understand code where possible?
In particular:
- The closure that's immediately called -- I understand you use it as a poor man's `try { }` block to allow `?` to work, but can't you use a function instead? Or at least a named closure?
- The overly complex expressions like `Some(quote! { }.into())` -- variables don't cost, and with a good name, they make code more readable

Reply:
You also [mentioned](https://github.com/godot-rust/gdext/pull/748#discussion_r1666972608) that this still clones, but where exactly? Or you mean just as part of the `map/collect` calls?

Reply:
It clones inside `make_method_docs` and `make_signal_docs` and `constant` (huh i have to rename that).

Reply:
is this better?
```rust
functions
    .iter()
    .map(make_method_docs)
    .collect::<Option<String>>()
    .and_then(|methods| {
        signals
            .iter()
            .map(make_signal_docs)
            .collect::<Option<String>>()
            .and_then(|signals| {
                constants
                    .iter()
                    .map(|ConstDefinition { raw_constant: x }| x)
                    .map(make_constant_docs)
                    .collect::<Option<String>>()
                    .map(|constants| {
                        quote! {
                            docs: ::godot::docs::InherentImplDocs {
                                methods: #methods,
                                signals: #signals,
                                constants: #constants,
                            }.into()
                        }
                    })
            })
    });
    ```

Reply:
or maybe i should make a `try!` macro?

Reply:
alright the function looks pretty good

Reply:
Looks quite good in the new code, thanks!